tools/PI/DevHome.PI/Controls/AddToolControl.xaml.cs

```diff
@@ -79,77 +128,295 @@ private void OkButton_Click(object sender, RoutedEventArgs e)
-        OtherArgsTextBox.Text = string.Empty;
+        LaunchRadio.IsChecked = true;
+        ArgumentsTextBox.Text = string.Empty;
         IsPinnedToggleSwitch.IsOn = true;
+        _selectedApp = null;
     }
 
     private ExternalTool? GetCurrentToolDefinition()
     {
         if (string.IsNullOrEmpty(ToolNameTextBox.Text) || string.IsNullOrEmpty(ToolPathTextBox.Text))
         {
-            WindowHelper.ShowTimedMessageDialog(this, invalidToolInfo, messageCloseText);
+            WindowHelper.ShowTimedMessageDialog(this, _invalidToolInfo, _messageCloseText);
             return null;
         }
 
-        var argType = ExternalToolArgType.None;
-
-        if (HwndRadio.IsChecked ?? false)
+        var activationType = ToolActivationType.Launch;
+        if (ProtocolRadio.IsChecked ?? false)
         {
-            argType = ExternalToolArgType.Hwnd;
+            activationType = ToolActivationType.Protocol;
         }
-        else if (ProcessIdRadio.IsChecked ?? false)
+        else if (_selectedApp is not null && _selectedApp.IsMsix)
         {
-            argType = ExternalToolArgType.ProcessId;
+            activationType = ToolActivationType.Msix;
         }
 
         return new(
             ToolNameTextBox.Text,
             ToolPathTextBox.Text,
-            argType,
-            PrefixTextBox.Text ?? string.Empty,
-            OtherArgsTextBox.Text ?? string.Empty,
+            activationType,
+            ArgumentsTextBox.Text,
+            _selectedApp?.AppUserModelId ?? string.Empty,
+            _selectedApp?.IconFilePath ?? string.Empty,
             IsPinnedToggleSwitch.IsOn);
     }
 
-    private void UpdateSampleCommandline(object sender, TextChangedEventArgs e)
+    private void CancelButton_Click(object sender, RoutedEventArgs e)
+    {
+        ClearValues();
+    }
+
+    private async void AppListButton_Click(object sender, RoutedEventArgs e)
+    {
+        _allApps.Clear();
+        SortedApps.Clear();
+
+        _itemCount = GetShortcuts();
+        if (_itemCount == 0)
+        {
+            _log.Error("Error getting _shortcuts");
+        }
+
+        var packageCount = GetPackages();
+        if (packageCount == 0)
+        {
+            _log.Error("Error getting _packages");
+        }
+
+        _itemCount += packageCount;
+        if (_itemCount == 0)
+        {
+            _log.Error("Error getting list of installed apps");
+            return;
+        }
+
+        // We get most of the data on a background thread, which
+        // reports intermittent progress.
+        IsLoading = true;
+        var progress = new Progress<int>(percent =>
+        {
+            // Update the progress report.
+            LoadingProgressTextRing.Value = percent;
+        });
+
+        await ProcessItemsAsync(progress);
+
+        foreach (var app in _allApps)
+        {
+            SortedApps.Add(app);
+        }
+
+        IsLoading = false;
+    }
+
+    private int GetShortcuts()
```

Comment:
>GetShortcuts [](http://example.com/codeflow?start=16&length=12)

Just an observation.... lots of code in here that isn't directly related to the UI. I wonder if it makes sense to move the non-UI code to a helper class which (sometime in the future) would be easier to unit test. Not blocking for this PR, just a though.

Reply:
Yes, I wanted to get this missing functionality in place so we can exercise it, and you're right we have another issue filed to do an audit/sweep of the all-up UX implementation.